lib/translator/status.cc

```diff
@@ -0,0 +1,252 @@
+      result.sense_key = scsi_defs::SenseKey::kMediumError;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedPowerLoss:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      result.asc = scsi_defs::AdditionalSenseCode::kWarningPowerLossExpected;
+      result.ascq =
+          scsi_defs::AdditionalSenseCodeQualifier::kWarningPowerLossExpected;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInternalDeviceError:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kHardwareError;
+      result.asc = scsi_defs::AdditionalSenseCode::kInternalTargetFailure;
+      result.ascq =
+          scsi_defs::AdditionalSenseCodeQualifier::kInternalTargetFailure;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedByRequest:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedSqDeletion:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedFailedFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedMissingFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInvalidNamespaceOrFormat:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc =
+          scsi_defs::AdditionalSenseCode::kAccessDeniedInvalidLuIdentifier;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kAccessDeniedInvalidLuIdentifier;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kLbaOutOfRange:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc = scsi_defs::AdditionalSenseCode::kLbaOutOfRange;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::kLbaOutOfRange;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kNamespaceNotReady:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kNotReady;
+      result.asc = scsi_defs::AdditionalSenseCode::
+          kLogicalUnitNotReadyCauseNotReportable;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kLogicalUnitNotReadyCauseNotReportable;
+      break;
+    default:
+      DebugLog("No SCSI translation for NVMe status with code %#x\n",
+               static_cast<uint8_t>(sc));
+      return StatusCode::kNoTranslation;
```

Comment:
What is downstream code expected to do when this case (or a similar one below) arises, since no status/sense_key is supplied? Should we set some catch-all values so execution can proceed?

Reply:
That's a good concern! For NVMe, there is only one status code for success, so we know that all others are some type of failure. Therefore, we can technically set the `SenseKey`, `AdditionalSenseCode`, etc. values to be some generic failure. However, generic failure codes don't really exist. 

As it currently is,  when there is no NVMe to SCSI translation, the `translator::ScsiStatus` object will remain default constructed, in which case both `SenseKey` and `AdditionalSenseCode` will be `kNoSense`. In this case, downstream code, after encountering a `translator::StatusCode::kNoTranslation`, can still proceed with the execution and send the various `kNoSense` codes back to the SCSI driver